src/lib/coda_base/pending_coinbase.ml

```diff
@@ -236,14 +236,41 @@ module T = struct
         (Merkle_tree.get_req ~depth (Hash.var_to_hash_packed t) addr)
         reraise_merkle_requests
 
-    let%snarkydef add_coinbase t coinbase =
+    (*TODO: change the argument (pk, amount) back to Coinbase_data.var after merging the coinbase-amount-to-unsigned pr *)
+    let%snarkydef add_coinbase t (pk, amount) =
       let%bind addr =
         request_witness Address.typ
           As_prover.(map (return ()) ~f:(fun _ -> Find_index_of_newest_stack))
       in
+      let equal_to_zero x = Amount.(equal_var x (var_of_t zero)) in
+      let chain if_ b ~then_ ~else_ =
+        let%bind then_ = then_ and else_ = else_ in
+        if_ b ~then_ ~else_
+      in
       handle
         (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
-           ~f:(fun stack -> Coinbase_stack.Stack.Checked.push stack coinbase ))
+           ~f:(fun stack ->
+             let total_coinbase_amount =
+               Currency.Amount.var_of_t Protocols.Coda_praos.coinbase_amount
+             in
+             let%bind rem_amount, `Underflow underflowed =
+               Currency.Amount.Checked.sub_flagged total_coinbase_amount amount
+             in
+             let%bind () = Boolean.Assert.is_true (Boolean.not underflowed) in
+             let%bind amount1_equal_to_zero = equal_to_zero amount in
+             let%bind amount2_equal_to_zero = equal_to_zero rem_amount in
+             let%bind stack_with_amount1 =
+               Coinbase_stack.Stack.Checked.push stack
+                 (pk, Amount.Signed.Checked.of_unsigned amount)
```

Comment:
There is an opportunity to save a lot of hashing here. Hashing a public key is about 1000 constraints and that could be shared between the two pushes. Maybe make an issue with an 'optimization' label